### PR TITLE
[wsu] Configure CNI on bootstrapped nodes

### PR DESCRIPTION
**Pull request by sebsoto**
This commit run the wmcb's configure-cni command on the bootstrapped
nodes. In addition it distributes the files needed to run the command to
the nodes.

This commit also adds an e2e test to the wsu test suite, which deploys
Windows and Linux pods, and ensures they can communicate between each
other.

### CODE REVIEW

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 1]
```diff
@@ -288,3 +298,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+// pullDockerImage pulls the designated image on the remote host
+func pullDockerImage(name string) error {
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+	command := "docker pull " + name
+	errorCode, err := winRMClient.Run(command, stdout, stderr)
+	if err != nil {
+		return fmt.Errorf("failed to remotely run docker pull: %s", err)
+	}
+	stderrString := stdout.String()
```
Comment (sebsoto) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: This is a mistake, ill change it once the first review pass is done

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 2]
```diff
@@ -288,3 +298,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+	assert.NoError(t, err, "Could not curl the Windows server from a Windows container")
+
+	// TODO: test Windows <-> Windows on different node
+}
+
+// waitUntilJobSucceeds will return an error if the job fails or reaches a timeout
+func waitUntilJobSucceeds(name string) error {
+	var job *batchv1.Job
+	var err error
+	for i := 0; i < 10; i++ {
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: You have code like this scattered around. Please make constants for you retries and use it everywhere.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 3]
```diff
@@ -288,3 +298,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+		if err != nil {
+			return err
+		}
+		if job.Status.Succeeded > 0 {
+			return nil
+		}
+		if job.Status.Failed > 0 {
+			return fmt.Errorf("job %v failed", job)
+		}
+		time.Sleep(5 * time.Second)
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Constant for wait time and use it everywhere.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 4]
```diff
@@ -288,3 +298,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+			}
+		}
+		time.Sleep(5 * time.Second)
+	}
+	return pod, fmt.Errorf("timed out waiting for pod %v to have ready container", pod)
+}
+
+// getPodWithLabelSelector returns the pod that matches the label selector. If more than one pod match the selector,
+// the function will return an error
+func getPodWithLabelSelector(selector metav1.LabelSelector) (v1.Pod, error) {
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: If you check if the deployment has scaled up then you won't need this function.

File: tools/ansible/tasks/wsu/main.yaml [hunk 5]
```diff
@@ -151,3 +162,29 @@
...
+
+    - name: Move the CNI folder to the install directory
+      win_copy:
+        src: "{{ win_temp_dir.path }}\\cni"
+        dest: "C:\\k\\"
+        remote_src: yes
+
+    - name: Create cni config
+      win_copy:
+        content: "{\"cniVersion\":\"0.2.0\",\"name\":\"OpenShiftNetwork\",\"type\":\"win-overlay\",\"capabilities\":{\"dns\":true},\"ipam\":{\"type\":\"host-local\",\"subnet\":\"{{ hybrid_subnet.stdout }}\"},\"policies\":[{\"name\":\"Endpoin
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply tools/ansible/tasks/wsu/main.yaml]: There is more to WINC-130 than what I am doing here. I am still hardcoding a value specified in that ticket. The value I'm generating dynamically needs to be done dynamically or this ticket wouldnt be possible

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 6]
```diff
@@ -288,3 +298,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+func pullDockerImage(name string) error {
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+	command := "docker pull " + name
+	errorCode, err := winRMClient.Run(command, stdout, stderr)
+	if err != nil {
+		return fmt.Errorf("failed to remotely run docker pull: %s", err)
+	}
+	stderrString := stdout.String()
+	if errorCode != 0 {
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: In this case it seems like it is. Whenever I have seen a failure it is reflected in the error code.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 4 above)
Comment (sebsoto) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: I still need the function to retrieve the pod IP

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 7]
```diff
@@ -74,6 +74,11 @@ var (
 	windowsServerImage = "mcr.microsoft.com/windows/servercore:ltsc2019"
 	// ubi8Image is the name/location of the linux image we will use for testing
 	ubi8Image = "registry.access.redhat.com/ubi8/ubi:latest"
+
+	// retryCount is the amount of times we will retry an api operation
+	retryCount = 10
+	// retryPeriod is the interval of time until we retry after a failure
+	retryPeriod = 5 * time.Second
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: `retryInterval`?

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (aravindhp) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Please call that out in a comment

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 8]
```diff
@@ -334,8 +334,14 @@ func testEastWestNetworking(t *testing.T) {
 	winServerDeployment, err := createWindowsServerDeployment("win-server", winServerCommand)
 	require.NoError(t, err, "Could not create Windows deployment")
 	defer deleteDeployment(winServerDeployment.Name)
+
 	// Wait until the server is ready to be queried
-	winServerPod, err := waitUntilPodHasRunningContainer(*winServerDeployment.Spec.Selector)
+	err = waitUntilDeploymentScaled(winServerDeployment.Name)
+	require.NoError(t, err, "Deployment was unable to scale")
+
+
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Extra line

File: tools/ansible/tasks/wsu/main.yaml [hunk 9]
```diff
@@ -151,3 +162,23 @@
...
+    - name: Get hybrid overlay subnet
+      delegate_to: localhost
+      shell: oc get nodes "{{ node_name.stdout }}" -o=jsonpath='{.metadata.annotations.k8s\.ovn\.org/hybrid-overlay-hostsubnet}'
+      register: hybrid_subnet
+
+    - name: Create cni config
+      win_copy:
+        content: "{\"cniVersion\":\"0.2.0\",\"name\":\"OpenShiftNetwork\",\"type\":\"win-overlay\",\"capabilities\":{\"dns\":true},\"ipam\":{\"type\":\"host-local\",\"subnet\":\"{{ hybrid_subnet.stdout }}\"},\"policies\":[{\"name\":\"Endpoin
+tPolicy\",\"value\":{\"Type\":\"OutBoundNAT\",\"ExceptionList\":[\"10.132.0.0/16\"]}},{\"name\":\"EndpointPolicy\",\"value\":{\"Type\":\"ROUTE\",\"DestinationPrefix\":\"10.132.0.0/16\",\"NeedEncap\":true}}]}"
+        dest: "{{ win_temp_dir.path }}\\cni.conf"
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/main.yaml]: There will be a double copy of `cni.conf` now. One will be present in `c:\k\cni\` and another in `c:\k\cni\config`. Please put this in `{{ win_temp_dir.path }}\\cni\config\cni.conf`...

File: tools/ansible/tasks/wsu/main.yaml [hunk 10]
```diff
@@ -151,3 +162,23 @@
...
+
+    - name: Configure CNI
+      win_shell: "{{ win_temp_dir.path }}\\wmcb.exe configure-cni --cni-dir=\"{{ win_temp_dir.path }\\cni\" --cni-config=\"{{ win_temp_dir.path }}\\cni.conf\""
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/main.yaml]: ...and reflect it here.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 4 above)
Comment (aravindhp) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Then that is the only thing this function should return.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 11]
```diff
@@ -401,24 +400,24 @@ func waitUntilDeploymentScaled(name string) error {
...
 	}
 	return fmt.Errorf("timed out waiting for deployment %v to scale", deployment)
 }
 
-// getPodWithLabelSelector returns the pod that matches the label selector. If more than one pod match the selector,
-// the function will return an error
-func getPodWithLabelSelector(selector metav1.LabelSelector) (v1.Pod, error) {
+// getPodIPWithLabelSelector returns the IP of the pod that matches the label selector. If more than one pod match the
+// selector, the function will return an error
+func getPodIPWithLabelSelector(selector metav1.LabelSelector) (string, error) {
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: `getPodIP` is sufficient as per go naming standards. The input implies that you are getting the pod IP based on the label selector.

File: tools/ansible/tasks/wsu/main.yaml [hunk 12]
```diff
@@ -151,3 +162,27 @@
       until: '"k8s.ovn.org/hybrid-overlay-distributed-router-gateway-mac" in overlay.stdout'
       retries: 12
       delay: 5
+
+    - name: Get hybrid overlay subnet
+      delegate_to: localhost
+      shell: oc get nodes "{{ node_name.stdout }}" -o=jsonpath='{.metadata.annotations.k8s\.ovn\.org/hybrid-overlay-hostsubnet}'
+      register: hybrid_subnet
+
+    - name: Create cni config directory
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/wsu/main.yaml]: I am trying to address this here - https://github.com/openshift/windows-machine-config-operator/pull/109

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 13]
```diff
@@ -288,3 +303,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+// testEastWestNetworking deploys Windows and Linux pods, and tests that the pods can communicate
+func testEastWestNetworking(t *testing.T) {
+	// Preload the image that will be used on the Windows node, to prevent download timeouts
+	// and separate possible failure conditions into multiple operations
+	err := pullDockerImage(windowsServerImage)
+	require.NoError(t, err, "Could not pull Windows Server image")
+
+	// This will run a Server on the container, which can be reached with a GET request
+	winServerCommand := []string{"powershell.exe", "-command",
+		"<#code used from https://gist.github.com/wagnerandrade/5424431#> ; $$listener = " +
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Any specific reason for using this command espcially reference to `wagnerandrade..`?

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 14]
```diff
@@ -288,3 +303,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+	// This will install curl and then curl the windows server.
+	linuxCurlerCommand := []string{"bash", "-c", "yum update; yum install curl -y; curl " + winServerIP}
+	linuxCurlerJob, err := createLinuxJob("linux-curler", linuxCurlerCommand)
+	require.NoError(t, err, "Could not create Linux job")
+	defer deleteJob(linuxCurlerJob.Name)
+	err = waitUntilJobSucceeds(linuxCurlerJob.Name)
+	assert.NoError(t, err, "Could not curl the Windows server from a linux container")
+
+	// test Windows <-> Windows on same node
+	// This will continually try to read from the Windows Server. We have to try multiple times as the network is flaky
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: If there is a known bug around this call this out and explain that we're using loop to try multiple times

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 15]
```diff
@@ -288,3 +303,246 @@ func testHNSNetworksCreated(t *testing.T) {
...
+	// on container startup
+	winCurlerCommand := []string{"powershell.exe", "-command", "for (($i =0), ($j = 0); $i -lt 10; $i++) { " +
+		"$response = Invoke-Webrequest -UseBasicParsing -Uri " + winServerIP +
+		"; $code = $response.StatusCode; echo \"GET returned code $code\";" +
+		"If ($code -eq 200) {exit 0}; Start-Sleep -s 10;}; exit 1" + winServerIP}
+	winCurlerJob, err := createWindowsServerJob("win-curler", winCurlerCommand)
+	require.NoError(t, err, "Could not create Windows job")
+	defer deleteJob(winCurlerJob.Name)
+	err = waitUntilJobSucceeds(winCurlerJob.Name)
+	assert.NoError(t, err, "Could not curl the Windows server from a Windows container")
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: nit: Webserver in the windows pod from another Windows container?

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 14 above)
Comment (sebsoto) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Its not a known bug, its something that I've found when testing and may just be the normal behavior of Windows containers

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 13 above)
Comment (sebsoto) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: This command was what @aravindhp used and had hosted for the kubecon demo

Comment (aravindhp) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: That shouldn't be the reason to use that here. What we did for kubecon was hack something together just to get it to work.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 14 above)
Comment (aravindhp) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: I don't think it is the network being flaky but us trying to run the tests before the container has finished its initial setup. That is what needs to be called out here.

**Pull request merged**